src/bind-all.ts

```diff
@@ -35,254 +35,31 @@ function getBinding(original: Binding, sharedOptions?: boolean | AddEventListene
   return binding;
 }
 
-export function bindAll<Target extends EventTarget, Type extends string>(
-  target: Target,
-  bindings: [Binding<Target, Type>],
-  sharedOptions?: boolean | AddEventListenerOptions,
-): UnbindFn;
-export function bindAll<Target extends EventTarget, Type1 extends string, Type2 extends string>(
-  target: Target,
-  bindings: [Binding<Target, Type1>, Binding<Target, Type2>],
-  sharedOptions?: boolean | AddEventListenerOptions,
-): UnbindFn;
-export function bindAll<
-  Target extends EventTarget,
-  Type1 extends string,
-  Type2 extends string,
-  Type3 extends string,
->(
-  target: Target,
-  bindings: [Binding<Target, Type1>, Binding<Target, Type2>, Binding<Target, Type3>],
-  sharedOptions?: boolean | AddEventListenerOptions,
-): UnbindFn;
-export function bindAll<
-  Target extends EventTarget,
-  Type1 extends string,
-  Type2 extends string,
-  Type3 extends string,
-  Type4 extends string,
->(
-  target: Target,
-  bindings: [
-    Binding<Target, Type1>,
-    Binding<Target, Type2>,
-    Binding<Target, Type3>,
-    Binding<Target, Type4>,
-  ],
-  sharedOptions?: boolean | AddEventListenerOptions,
-): UnbindFn;
-export function bindAll<
-  Target extends EventTarget,
-  Type1 extends string,
-  Type2 extends string,
-  Type3 extends string,
-  Type4 extends string,
-  Type5 extends string,
->(
-  target: Target,
-  bindings: [
-    Binding<Target, Type1>,
-    Binding<Target, Type2>,
-    Binding<Target, Type3>,
-    Binding<Target, Type4>,
-    Binding<Target, Type5>,
-  ],
-  sharedOptions?: boolean | AddEventListenerOptions,
-): UnbindFn;
-export function bindAll<
-  Target extends EventTarget,
-  Type1 extends string,
-  Type2 extends string,
-  Type3 extends string,
-  Type4 extends string,
-  Type5 extends string,
-  Type6 extends string,
->(
-  target: Target,
-  bindings: [
-    Binding<Target, Type1>,
-    Binding<Target, Type2>,
-    Binding<Target, Type3>,
-    Binding<Target, Type4>,
-    Binding<Target, Type5>,
-    Binding<Target, Type6>,
-  ],
-  sharedOptions?: boolean | AddEventListenerOptions,
-): UnbindFn;
 export function bindAll<
-  Target extends EventTarget,
-  Type1 extends string,
-  Type2 extends string,
-  Type3 extends string,
-  Type4 extends string,
-  Type5 extends string,
-  Type6 extends string,
-  Type7 extends string,
+  TTarget extends EventTarget,
+  TTypes extends ReadonlyArray<InferEventType<TTarget> & string>,
 >(
-  target: Target,
+  target: TTarget,
   bindings: [
-    Binding<Target, Type1>,
-    Binding<Target, Type2>,
-    Binding<Target, Type3>,
-    Binding<Target, Type4>,
-    Binding<Target, Type5>,
-    Binding<Target, Type6>,
-    Binding<Target, Type7>,
+    ...{
+      [K in keyof TTypes]: {
+        type: TTypes[K] | (string & {});
+        listener: Listener<
+          TTarget,
+          InferEvent<
+            TTarget,
+            // `& string` "cast" is not needed since TS 4.7 (but the repo is using TS 4.6 atm)
```

Comment:
We can update to TS@latest to remove this

Reply:
the question is - what should be the minimum TS requirement for the end users of the package?

Reply:
4.7 has been released smth like 8 months ago. DefinitelyTyped officially supports the last 2 years of TS releases - that doesn't necessarily mean that you have to have the same policy though. It's not exactly like you have millions of downloads of this package - so I think that it's reasonable to prioritize your own use cases, ease of maintenance, etc.

That being said - this cast here is completely harmless. I just put this comment here to give a little bit of context behind this cast and also to leave the information on when it could be removed in the future (when you drop support for TS 4.6).

[The more important change](https://github.com/microsoft/TypeScript/pull/51770) for this lib is coming in TS 5.0 - but the difference between the version of the code in this PR and the potential version of the code that would be compatible with TS 5.0 is also really subtle and you don't even have to follow up on that one immediately either.

However, it still might be worthwhile to act on it when TS 5.0 gets released (March 14th). That would remove the autocomplete capabilities for <5.0 users but the types would become *slightly* better (conceptually and all). So this is kinda up to you - I think that it's a fair game that autocompletion would behave differently between those TS versions. The code/types themselves would still work - the only difference would be the DX in the IDE.

Reply:
To confirm my understanding:

- The current version would work on TS 4.6+ (and 5.0)
- After 5.0 is released we could do some internal cleanup of the code and still have good auto complete

Questions:
- After 5.0 could we keep the currently proposed change and maintain good auto-complete for everybody? (Could eventually move to 5.0 later)

cc @Andarist

Reply:
> The current version would work on TS 4.6+ (and 5.0)

Yes.

> After 5.0 is released we could do some internal cleanup of the code and still have good auto complete

Yes, but if we make that cleanup then TS 4.x would lose autocomplete

> After 5.0 could we keep the currently proposed change and maintain good auto-complete for everybody? (Could eventually move to 5.0 later)

Yes, we don't have to do this cleanup immediately. We are free to leave it as is for a foreseeable future. We could also introduce said the cleanup, copy the current types into a separate file, and use `typesVersions ` to point TS 4.x users to that other file. This arguably complicated the setup and stuff - so it's up to you if you would be comfortable with this or not